### DISCUSSION TITLE
Is it possible to make Mutable Types support subtype?

### DISCUSSION THREAD

**Discussion by NoahStoryM** [racket/typed-racket#1129] [Ideas]
I tried to write mlength for mlist, but it can't work for `(MPairof Natural Null)` type:
```
(: mlength (-> (MListof Natural) Natural))
(define mlength
  (λ (m)
    (if (null? m)
        0
        (add1 (mlength (mcdr m))))))

(: mls (MPairof Natural Null))
(define mls (mcons 9 '()))


(mlength mls)
```

I can understand it: `(MPairof Natural Null)` is not subtype of `(MListof Natural)`
because we can't run `(set-mcdr! mls (ann (mcons 8 '()) (MListof Natural)))`.
But it's legal for mlength to run  `(set-mcdr! m (ann (mcons 8 '()) (MListof Natural)))`
in its body when mls bound to m, which cause conflict.

I wonder if it's possible to make Mutable Types support subtype, and when TR 
found conflict, just throw an exception?
(I haven't studied type theory, I just wonder if it's feasible?)

for example:
```
(: mls (MListof Natural))
(define mls (mcons 9 '()))

(: change-1st! (-> (MListof Any) #:immediately Void))
(define change-1st!
  (λ (mls)
    (unless (null? mls)
      (set-mcar! mls 'sym))))

(displayln mls)
(change-1st! mls)
(displayln mls)
```

when apply mls to change-1st!, it will record mls's type and value.
If mls's type is changed, then TR throws an exception.
(typecheck happens when the Mutable Type's assign operator was applied if the keyword is #:immediately,
or it happens when the calculation is finished if the keyword is #:end,
or it doesn't happen if the keyword is #:never)

**Comment by samth** [comment]
This is a possible language design -- it's roughly how the array type in Java works. When you create an array in Java, it records the type it's supposed to have, and if you try to assign something wrong to it then there's a runtime error. So we could have a design where mutable pairs recorded the types of their two fields when they were created, and did runtime checks for those types on assignment. Then we would have covariant subtyping on mutable pairs.

Unfortunately, this is a very different design than what Racket has today, so it would make it impossible for mpairs in Typed Racket to be the same as mpairs in Racket. That's a reasonable choice for a language, but Typed Racket aims to be highly-compatibly with Racket, so it won't work for Typed Racket.

I'm going to move this to a "discussion" because that seems like a better place for discussing possible options for programs like this.

**Reply by NoahStoryM** [reply]
I see. In addition, I am a little curious that there are many restrictions on the use of mutable types in TR (for example, it is impossible to define predicate for them). Are these restrictions mainly due to subtypes?

**Reply by samth** [reply]
These are indeed related issues. Consider this program:

```
(define-predicate p? (MPair Integer Integer))
(define v : (MPair Any Any) (mcons 1 2))

(if (p? v) (+ 1 (mcar v)) 2)
```

Should this type check? Unfortunately, it can't, because in between the `(p? v)` and the `(mcar v)`, someone else could do `(set-mcar! v 'x)` and then we'd have a type error. 

The fundamental issue is that subtyping (or rather, covariance) is possible only when _your code_ is making that choice of the subtype. But when you have a channel of communication from _someone else_, as in a function result or reading the value of a mutable location, then that someone else is making the choice of the subtype, so you can't know which one it's going to be.

**Comment by AlexKnauth** [comment]
If every type-argument in a mutable type constructor were replaced with a pair of types with opposite variance, we could get something. I'll use `(Boxof t)` as an example. If we replace that `t` with a pair of types, `t-` and `t+`, then we get subtyping like this:
```racket
b- <: a-     and     a+ <: b+
------------------------------
(Boxof a- a+) <: (Boxof b- b+)
```
In other words in `(Boxof t- t+)`, `t-` is contravariant while `t+` is covariant. The type `t-` represents the types that may be written into the box with `set-box!`, and the type `t+` represents the types that may be read from the box with `unbox`.

The [`box/c`](https://docs.racket-lang.org/reference/data-structure-contracts.html#%28def._%28%28lib._racket%2Fcontract%2Fprivate%2Fbox..rkt%29._box%2Fc%29%29) contract combinator already supports this:
```racket
> (contract-stronger? (box/c exact-integer? exact-integer?)
                      (box/c exact-integer? any/c))
#t
> (contract-stronger? (box/c exact-integer? exact-integer?)
                      (box/c any/c any/c))
#f
```

------------

For the example above with `MPairof`, no, `(MPairof Natural Null)` should still not be a subtype of `(MListof Natural)`, but replacing each type with a pair of types written as `(t-, t+)`, would give:
```racket
(MPairof (Natural, Natural) ((MListof (Natural, Natural)), Null)) <: (MListof (Natural, Natural))
```

This would also allow a more expressive type for `mlength` that says that it can't mutate the mlist its given... in other words it would allow read-only versions of mutable types:
```racket
(define-type (ReadOnlyMListof t)
  (U Null (MPairof (Nothing, t) (Nothing, (ReadOnlyMListof t)))))

(: mlength (-> (ReadOnlyMListof Natural) Natural))
```
And then with that more expressive type the new `mlength` can be called with an `(MPairof Natural Null)` with no problem because:
```racket
(MPairof (Natural, Natural) (Null, Null)) <: (MPairof (Nothing, Natural) (Nothing, (ReadOnlyMListof Natural)))
--------------------------------------------------------------------------------------------------------------
(MPairof (Natural, Natural) (Null, Null)) <: (ReadOnlyMListof Natural)
```

**Reply by AlexKnauth** [reply]
I started work on this in https://github.com/racket/typed-racket/pull/225, but never finished it, and now a few years later there are merge conflicts.

And that's only for `Boxof` types... would probably need a more generalized way of dealing with the opposite-variance-type-pairs for types such as `MPairof` that have multiple type-arguments. It seems "ugly" to me to just put in 4 type-arguments with no extra syntax making it obvious that they come it opposite-variance pairs.

**Reply by AlexKnauth** [reply]
I'm considering an MLsub-inspired syntax like this for opposite-variance-type-pairs:
```racket
(Boxof Natural) = (Boxof (- Natural + Natural))
(Boxof (+ Natural)) = (Boxof (- Nothing + Natural))
(Boxof (- Natural)) = (Boxof (- Natural + Any))
```
Where in `(Boxof (- a + b))`, the `a` is the contravariant one and the `b` is the covariant one.

And so a `ReadOnly` version will use `+` a lot to indicate that it only uses the covariant position:
```racket
(define-type (ReadOnlyMListof t)
  (U Null (MPairof (+ t) (+ (ReadOnlyMListof t)))))
```

**Reply by NoahStoryM** [reply]
This is like Parameter. I have a doubt about Parameter: I regard parameter as a box with dynamic scope.
But what makes me feel strange is that racket doesn't have Immutable Parameter?
(I think Immutable Parameter can only get different values through parameterize, but not directly modify its value.

For example:
```
(: p1 (Immutable-Parameter Number))
(define p1 (make-immutable-parameter 1))

;; can't run (p1 2)
(p1)

(parameterize ([p1 2])
  ;; can't run (p1 2)
  (p1))

;; can't run (p1 2)
(p1)

```
).

I tried to use `(Parameter Nothing Number)` to define an Immutable value before. And now I know definition like this
is read-only, not Immutable. But I am still confused how to understand the type relationship between a parameter variable and it's use in parameterize? I feel the implementation of parameterize is somewhat similar to recording the old value and changing back to the old value at end, so it relies on assignment. But dynamic scope seems to have nothing to do with assignment, it is just another option different from lexical scope?

**Reply by NoahStoryM** [reply]
Although (Parameter Nothing Number) is read only, not immutable.
But should we regard the binding in the parameterize is just a kind of assignment?
```
(: p1 (Parameter Nothing Number))
(define p1 (make-parameter 1))

;; should't run (p1 2)
(p1)

(parameterize ([p1 2])
  ;; should't run (p1 2)
  (p1))

;; should't run (p1 2)
(p1)
```

**Reply by AlexKnauth** [reply]
The `t-` in `(Parameter t- t+)` needs to account for all places where any operation takes it as an input to put into the data, including both `parameterize` and `(p1 x)`. It isn't about "mutability", it's about whether a type is used for inputs and outputs in the functions that access it.

**Reply by NoahStoryM** [reply]
So (Parameter W R) is more similar to (case-> (-> W Void) (-> R)), 
which is different from a box with dynamic scope? When it needs input,
the input's type must be W?


But if there are a type like dynamic scope box, and an operator dynamic-box, 
should the binding type be W or R?

For example:
```
(: b1 (Dynamic-Immutable-Boxof Natural))
(define b1 (dynamic-immutable-box 1))

(: b2 (Dynamic-Boxof Natural Number))
(define b2 (dynamic-box -1))

(unbox b1) ; 1
(unbox b2) ; -1

(dynamic-box ([b1 0]
              [b2 val]  ; should val be W type (Natural) or R type (Number)?
              )
  (unbox b1) ; 0
  (unbox b2) ; val
  
  (set-box! b2 2)
  (unbox b2) ; 2
  )

(unbox b1) ; 1
(unbox b2) ; -1

```

**Reply by NoahStoryM** [reply]
> So (Parameter W R) is more similar to (case-> (-> W Void) (-> R)),
> which is different from a box with dynamic scope? When it needs input,
> the input's type must be W?
> 
> But if there are a type like dynamic scope box, and an operator dynamic-box,
> should the binding type be W or R?
> 
> For example:
> 
> ```
> (: b1 (Dynamic-Immutable-Boxof Natural))
> (define b1 (dynamic-immutable-box 1))
> 
> (: b2 (Dynamic-Boxof Natural Number))
> (define b2 (dynamic-box -1))
> 
> (unbox b1) ; 1
> (unbox b2) ; -1
> 
> (dynamic-box ([b1 0]
>               [b2 val]  ; should val be W type (Natural) or R type (Number)?
>               )
>   (unbox b1) ; 0
>   (unbox b2) ; val
>   
>   (set-box! b2 2)
>   (unbox b2) ; 2
>   )
> 
> (unbox b1) ; 1
> (unbox b2) ; -1
> ```

I guess I know where I got it wrong:

- From racket guide:
The parameterize form is not a binding form like let; each use of location above refers directly to the original definition. A parameterize form adjusts the value of a parameter during the whole time that the parameterize body is evaluated, even for uses of the parameter that are textually outside of the parameterize body:

parameterize is not similar to let, dynamic-box, it does not create a new reference:
```
(: p1 (Parameter Natural Number))
(define p1 (make-parameter -1))

(define p2 p1)

(parameterize ([p1 2]); adjusts the value of p1
   (displayln (eq? p1 p2)); #t
   )
```

```
(: b1 (Dynamic-Immutable-Boxof Natural))
(define b1 (dynamic-immutable-box 1))
(define b2 b1)

(unbox b1); 1
(unbox b2); 1

(dynamic-box ([b1 0]); create a new dynamic-immutable-box.
   (displayln (eq? b1 b2)); #f
   )
```

**Reply by AlexKnauth** [reply]
Would you have better understanding if we labeled them "Input" and "Output" instead of "Write" and "Read"? In `(Boxof t- t+)`, the `t-` is the type used inputs, and `t+` is the type used for outputs.

Or, since it's already using `-` and `+`, would you understand it better if we labeled them "Negative" and "Positive"? This is related to negative and positive blame for values crossing a contract boundary, if you're familiar with that.

**Reply by NoahStoryM** [reply]
I guess I could understand this. In my understanding, ~~all operations on a subtype can also be performed on its supertype~~ all operations on a supertype can also be performed on its subtype.
The two most typical operations are read and write. The supertype can safely read "more" types and write "fewer" types than its subtype.

So
```
a+ <: b+
------------------------------
(Boxof Nothing a+) <: (Boxof Nothing b+)
```

```
b- <: a-
------------------------------
(Boxof a- Nothing) <: (Boxof b- Nothing)
```

**Reply by AlexKnauth** [reply]
> all operations on a subtype can also be performed on its supertype

Not quite, you have this part backwards: all operations on a supertype can also be performed on its subtype, so if you have `(f b)` where `b : B`, and `A <: B`, then `(f a)` should also typecheck where `a : A`. All operations on the supertype `B` can also be performed on its subtype `A`.

**Reply by NoahStoryM** [reply]
Thank you very much! I confused subtype and supertype before. (f a) and (f b) example is very easy to understand!

**Comment by samth** [comment]
In general, to look for more work on this topic, searching for "declaration site variance" vs "use-site variance" is the way to find papers, eg: https://dl.acm.org/doi/abs/10.1145/2489837.2489842

**Comment by NoahStoryM** [comment]
It seems that supporting write types is more difficult than contracts: https://racket.discourse.group/t/a-new-way-to-type-mutable-data-structures-with-opposite-variance-in-typed-racket/1846/7?u=noahstorym .
